### ISSUE TITLE
Slow loading times of BO page "Price Catalog Rules" due to full smarty cache clear at every load

### ISSUE THREAD

**Issue by gennaris** [PrestaShop/PrestaShop#29298]
### Prerequisites

- [X] I understand and accept the project's [code of conduct](https://github.com/PrestaShop/PrestaShop/blob/develop/CODE_OF_CONDUCT.md)
- [X] I have already [searched in existing issues](https://github.com/PrestaShop/PrestaShop/issues?q=is%3Aissue+label%3ABug) and found no previous report of this bug.

### Describe the bug and add screenshots

If you have a lot of smarty cache , 
backoffice "Price Catalog Rules" page loading time can take a lot of time or timeout the page due to **Tools::clearSmartyCache();**  being called :

```
    public function postProcess()
    {
        Tools::clearSmartyCache();

        return parent::postProcess();
    }
```

in:  **controllers/admin/AdminSpecificPriceRuleController.php**

This method is called during page load and / or when submitting modifications.

This sounds nosense to me, I don't see a good reason to clear the smarty cache when loading a BO controller.

### Expected behavior

Unless I'm missing something, BO Page loading should not be affected by front office smarty cache clear

### Steps to reproduce

1. Fill your shop with a lot of smarty cache (difficult part 😀)
2. Go on **Catalog > Discounts > Catalog Price Rules**
3. See the operation takes a lot of time as the folders  **var/cache/prod/smarty/cache** and  **var/cache/prod/smarty/compile** are getting totally wiped before the load of the page.

### PrestaShop version(s) where the bug happened

1.7.8.7

### PHP version(s) where the bug happened

7.4

**Comment by Hlavtox** [comment]
It was added in this commit, but god knows why :D https://github.com/PrestaShop/PrestaShop/commit/2556d64f6e1cfd636dc4152dadc7ff61c5433cd7

**Comment by AureRita** [comment]
Hi @gennaris 

Thank for your report,
This issue is too technical for me, ping @PrestaShop/prestashop-core-developers, can someone please reproduce this issue ? 

Thanks

**Comment by ghost** [comment]
Hello @AureRita 


https://user-images.githubusercontent.com/5064590/184079998-28162ce2-98a6-41b3-a1f2-6af26da283dc.mp4

**Comment by AureRita** [comment]
Thank you @okom3pom ,

Now I see where the issue come. For me, yes, we have to withdraw this part of code, because it haven't to be here, but currently I'm not a developper and I don't know if it correct maybe some issue with cart rules. So we just have to wait them, sorry

**Comment by ghost** [comment]
We could replace it by :arrow_down:  to keep the cache removal because _cart rules is cart rules :rofl:_ 

```
    public function processAdd()
    {
        Tools::clearSmartyCache();

        return parent::processAdd();
    }

    public function processUpdate()
    {
        Tools::clearSmartyCache();

        return parent::processUpdate();
    }  
```

let's wait for a dev :)

**Comment by ghost** [comment]
@AureRita good news

I asked if a committer could validate an issue with the label `Waiting for dev`

It's ok

**Comment by FabienPapet** [comment]
I think the the cache "needs" to be cleared for frontend if you add / update / remove rules to avoid issue with outdated data for customers.

@okom3pom 's solution looks ok for me (did not tested), if we handle correctly add / update / delete, then I think this code can be removed from `postProcess`

**Comment by gennaris** [comment]
@okom3pom solution only partially mitigates the problem, now the cache is not cleared when the page loads but it does after a cart rule save/update action, resulting in the same issue (in the case of my customer he generates GB of smarty cache every day).

From my verifications it does not make any sense to perform a **Tools::clearSmartyCache();** at all after a backoffice action,
the only change caused by a cart rule editing action are product prices, but even commenting out the line and applying a cart rule, I find that the product prices at front office are immediatly OK so I assume they are not cached - maybe back in the days of PS 1.4 (when commit was introduced) there was a need to do it for other unknown reasons ?

**Comment by ghost** [comment]
I agree, the possible impact is the cache of the modules ( ps_specials for example ) and normally they are hook on updateProduct to clear the cache. 

And when a specific price expires there is no cache cleaning :interrobang: 

But PrestaShop cart rules are on the dark side of the force.

**Comment by Gipielle** [comment]
Thank you @gennaris for report, i also have this problem in many ps versions

**Comment by jalarcon666** [comment]
As a temporary solution until PrestaShop applies the change to the CORE, change the following in **smarty.config.inc.php**:

$smarty->use_sub_dirs = true;

to

$smarty->use_sub_dirs = false;

Since Smarty 2.6.2 this value should be false by default but in the Presta CORE it has not been modified, so it is necessary to apply it manually.

Since syscalls are reduced, performance appears to be better.

PR: #40716

**Comment by gennaris** [comment]
A simple fix was already proposed in my PR #29592, but it looks like it didn’t really get proper attention/review and was eventually closed as stale.

I’ve been running that patch in production for quite a while on several large PrestaShop stores and it works reliably: no regressions, and it removes the huge performance hit caused by wiping the whole Smarty cache on BO load.

From real-world usage, this is a safe change and it solves the issue described here.

**Comment by boo-code** [comment]
Still exactly as reported on 9.2.x - `AdminSpecificPriceRuleController::postProcess()` line 375:

```php
public function postProcess()
{
    Tools::clearSmartyCache();

    return parent::postProcess();
}
```

and `postProcess()` runs on every request to the page, so opening the list wipes the front-office cache.

Looking at the thread, the disagreement was never really about whether the clear is wrong - it was about whether to keep it at all. @FabienPapet said it is needed when rules are added, updated or removed; @okom3pom proposed moving it onto the mutating actions; @gennaris argued it is unnecessary altogether and PR #29592 removed it outright, which was then closed unmerged in September 2023.

The middle option is small and does not need that argument settled. `$this->action` is set by `AdminController::initProcess()` before `postProcess()` runs, so the clear can simply be scoped to requests that dispatched an action:

```php
$return = parent::postProcess();

if (!empty($this->action) && !in_array($this->action, self::NON_MUTATING_ACTIONS, true)) {
    Tools::clearSmartyCache();
}

return $return;
```

with `NON_MUTATING_ACTIONS` being `new`, `view`, `details`, `export`, `reset_filters`. Listing, filtering, exporting and opening the form stop clearing anything; create, edit, toggle and delete still do. Bulk actions come through as `'bulk' . $action`, so they keep working without being enumerated.

It is a deny list on purpose. Missing a mutating action would show customers stale prices; missing a read-only one only costs a needless clear, which is what the page did on every request anyway - so the failure mode degrades to today's behaviour rather than to wrong prices. It also runs after `parent::postProcess()` now, so the cache is dropped once the rule has actually changed.

I have that ready against `9.2.x`. @gennaris - this keeps the invalidation FabienPapet asked for, which is the part that stalled your PR; if you would still rather see it gone entirely that is a separate discussion, but this at least stops the page load doing it.

**Comment by gennaris** [comment]
This issue has officially survived long enough to be reviewed by an AI agent. I wasn’t expecting that plot twist. 😄

I’m curious to see whether my 2022 implementation still makes sense against today’s codebase, or whether it breaks any assumptions introduced since then.

Good luck with the archaeology!
